src/core/bpf-firewall.c

```diff
@@ -863,61 +860,47 @@ int bpf_firewall_supported(void) {
+         * supported (which was added in kernel 4.15). We use a similar logic as before, but this time we use
+         * the BPF_PROG_ATTACH bpf() call and the BPF_F_ALLOW_MULTI flags value. Since the flags are checked
+         * early in the system call we'll get EINVAL if it's not supported, and EBADF as before if it is
+         * available. */
 
         zero(attr);
         attr.attach_type = BPF_CGROUP_INET_EGRESS;
         attr.target_fd = -1;
         attr.attach_bpf_fd = -1;
         attr.attach_flags = BPF_F_ALLOW_MULTI;
 
-        if (bpf(BPF_PROG_ATTACH, &attr, sizeof(attr)) < 0) {
-                if (errno == EBADF) {
-                        log_debug_errno(errno, "Got EBADF when using BPF_F_ALLOW_MULTI, which indicates it is supported. Yay!");
-                        return supported = BPF_FIREWALL_SUPPORTED_WITH_MULTI;
-                }
-
-                if (errno == EINVAL)
-                        log_debug_errno(errno, "Got EINVAL error when using BPF_F_ALLOW_MULTI, which indicates it's not supported.");
-                else
-                        log_debug_errno(errno, "Got unexpected error when using BPF_F_ALLOW_MULTI, assuming it's not supported: %m");
-
-                return supported = BPF_FIREWALL_SUPPORTED;
-        } else {
-                bpf_firewall_unsupported_reason =
-                        log_debug_errno(SYNTHETIC_ERRNO(EBADE),
-                                        "Wut? Kernel accepted our invalid BPF_PROG_ATTACH+BPF_F_ALLOW_MULTI call? "
-                                        "Something is weird, assuming BPF firewalling is broken and hence not supported.");
-                return supported = BPF_FIREWALL_UNSUPPORTED;
-        }
+        r = RET_NERRNO(bpf(BPF_PROG_ATTACH, &attr, sizeof attr));
+        if (r >= 0)
+                return log_unsupported(SYNTHETIC_ERRNO(EBADE),
+                                       "Wut? Kernel accepted our invalid BPF_PROG_ATTACH+BPF_F_ALLOW_MULTI call? "
+                                       "Something is weird, assuming BPF firewalling is broken and hence not supported.");
+        else if (r == -EBADF) {
+                log_debug_errno(r, "Got EBADF when using BPF_F_ALLOW_MULTI, which indicates it is supported. Yay!");
+                return supported = BPF_FIREWALL_SUPPORTED_WITH_MULTI;
+        } else if (r == -EINVAL)
+                log_debug_errno(r, "Got EINVAL error when using BPF_F_ALLOW_MULTI, which indicates it's not supported.");
+        else
+                log_debug_errno(r, "Got unexpected error when using BPF_F_ALLOW_MULTI, assuming it's not supported: %m");
+
+        return supported = BPF_FIREWALL_SUPPORTED;
 }
 
 void emit_bpf_firewall_warning(Unit *u) {
         static bool warned = false;
 
-        assert(u);
-        assert(u->manager);
-
-        if (!warned && !MANAGER_IS_TEST_RUN(u->manager)) {
+        if (!warned) {
```

Comment:
btw, following our usual rule of "exit early/minimize indentation", I think this check should actually be inverted, so that we return here immediately if we already warned.